_posts/2019-12-10-variational-gaussian-process/vgp.Rmd

```diff
@@ -0,0 +1,549 @@
+blast_furnace_slag:superplasticizer  -0.6838     1.2812  -0.534 0.593676    
+blast_furnace_slag:coarse_aggregate  -1.0592     0.6416  -1.651 0.099154 .  
+blast_furnace_slag:fine_aggregate     2.0579     0.5538   3.716 0.000217 ***
+blast_furnace_slag:age                4.7563     1.1148   4.266 2.23e-05 ***
+fly_ash:water                        -2.7131     0.9858  -2.752 0.006054 ** 
+fly_ash:superplasticizer             -2.6528     1.2553  -2.113 0.034891 *  
+fly_ash:coarse_aggregate              0.3323     0.7004   0.474 0.635305    
+fly_ash:fine_aggregate                2.6764     0.7817   3.424 0.000649 ***
+fly_ash:age                           7.5851     1.3570   5.589 3.14e-08 ***
+water:superplasticizer                1.3686     0.8704   1.572 0.116289    
+water:coarse_aggregate               -1.3399     0.5203  -2.575 0.010194 *  
+water:fine_aggregate                 -0.7061     0.5184  -1.362 0.173533    
+water:age                             0.3207     1.2991   0.247 0.805068    
+superplasticizer:coarse_aggregate     1.4526     0.9310   1.560 0.119125    
+superplasticizer:fine_aggregate       0.1022     1.1342   0.090 0.928239    
+superplasticizer:age                  1.9107     0.9491   2.013 0.044444 *  
+coarse_aggregate:fine_aggregate       1.3014     0.4750   2.740 0.006286 ** 
+coarse_aggregate:age                  0.7557     0.9342   0.809 0.418815    
+fine_aggregate:age                    3.4524     1.2165   2.838 0.004657 ** 
+---
+Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
+
+Residual standard error: 8.327 on 788 degrees of freedom
+Multiple R-squared:  0.7656,	Adjusted R-squared:  0.7549 
+F-statistic: 71.48 on 36 and 788 DF,  p-value: < 2.2e-16
+```
+
+We also store the predictions on the test set, for later comparison.
+
+```{r}
+linreg_preds1 <- fit1 %>% predict(test[, 1:8])
+linreg_preds2 <- fit2 %>% predict(test[, 1:8])
+
+compare <-
+  data.frame(
+    y_true = test$strength,
+    linreg_preds1 = linreg_preds1,
+    linreg_preds2 = linreg_preds2
+  )
+```
+
+With no further preprocessing required, the [tfdatasets](https://tensorflow.rstudio.com/guide/tfdatasets/introduction/) input pipeline ends up nice and short:
+
+```{r}
+create_dataset <- function(df, batch_size, shuffle = TRUE) {
+  
+  df <- as.matrix(df)
+  ds <-
+    tensor_slices_dataset(list(df[, 1:8], df[, 9, drop = FALSE]))
+  if (shuffle)
+    ds <- ds %>% dataset_shuffle(buffer_size = nrow(df))
+  ds %>%
+    dataset_batch(batch_size = batch_size)
+  
+}
+
+# just one possible choice for batch size ...
+batch_size <- 64
+train_ds <- create_dataset(train, batch_size = batch_size)
+test_ds <- create_dataset(test, batch_size = nrow(test), shuffle = FALSE)
+```
+
+And on to model creation. 
+
+## The model
+
+Model definition is short as well, although there are a few things to expand on. Don't execute this yet:
+
+```{r}
+model <- keras_model_sequential() %>%
+  layer_dense(units = 8,
```

Comment:
Why this dense layer?

Reply:
yeah, this is explained in 

## Discussion: Feature Relevance

but as a reader, I'd also want to know already, so maybe I should add a footnote here already...

Reply:
Gotcha. Are there examples in the literature (I'm not familiar with this topic, btw...) that does this? It feels a little weird to me since we're trying to fit the process to a moving target of X's, but maybe that's OK.

Reply:
yeah, - myself i didn't find that the article (Hensmans et al.) actually said so, but i may have missed it, so i guess i'm relying on the TFP guys' judgments here (that these should actually be learnable parameters)